### PR TITLE
OAuth2 flow (#1656)

### PR DESCRIPTION
**Pull request by stefanorosanelli**
This PR solves #1656 

## Main changes

 * new `TokenMiddleware` used to decode JWT payload and identify client applications using JWT payload or API KEY as fallback - this middleware is inserted _before_ `RoutingMiddleware`

 * new `JWTHandler` utility class used to encode/decode JWT tokens 

 * `JwtAuthenticate` class has been modified:
    - encode/decode payload operations have been removed, payload is available in a request attribute
    - `refresh_token` use case is explicitly handled 

 * `LoginController` class has been modified:
    - `client_credentials` grant type use case is handled 
    - `refresh_token` grant type is explicitly set in request data if missing
    - `refresh_token` use case when only client credentials are set is also handled (i.e. JWT contains only application data and has no `user` information, call is anonymous) 
    
 * a new `applications.client_secret` DB column and property has been introduced 
 
 * a new `findCredentials` finder in `ApplicationsTable` has been introduced in order to find applications using standard `client_id/client_secret` keys where `api_key` field is used as `client_id` 

 * a new `AuthenticationTest` integration test case has been added in order to show the OAuth2 flow implemented in this PR
  
## Notes

* some `JwtAuthenticate` configurations have been removed, namely  `'scope', 'finder', 'contain', 'passwordHasher',  'queryDatasource'` - they are obsolete or not used anymore

### CODE REVIEW

File: plugins/BEdita/API/src/Controller/LoginController.php [hunk 1]
```diff
@@ -126,6 +128,28 @@ public function login(): void
...
+    protected function setGrantType(): void
+    {
+        if (!empty($this->request->getData('grant_type'))) {
+            return;
+        }
+
+        $data = $this->request->getData();
+        if (empty($data)) {
+            $this->request = $this->request->withData('grant_type', 'refresh_token');
+        } elseif (!empty($data['username']) && !empty($data['passwrod'])) {
```
Comment (batopa) [inline plugins/BEdita/API/src/Controller/LoginController.php]: typo `passwrod`

File: plugins/BEdita/API/src/Utility/JWTHandler.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+class JWTHandler
+{
+    /**
+     * Decode JWT token.
+     *
+     * @param string $token JWT token to decode.
+     * @return array The token's payload as a PHP array.
+     * @throws \Exception Throws an exception if the token could not be decoded.
+     */
+    public static function decode(string $token)
```
Comment (batopa) [inline plugins/BEdita/API/src/Utility/JWTHandler.php]: I would add a method property `$options` that can be used to decode the `JWT`. for example
```php
public static function decode(string $token, array $options = []): array
{
    $options += [
        'key' => Security::getSalt(),
        'algorithms' => Configure::read('Security.jwt.algorithm') ?: 'HS256',
    ];
}

$payload = JWT::decode($token, $options['key'], (array)$options['algorithms']);
```

File: plugins/BEdita/API/src/Middleware/TokenMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next Callback to invoke the next middleware.
+     * @return \Psr\Http\Message\ResponseInterface A response
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
+    {
+        $payload = [];
+        $token = $this->getToken($request);
+        if (!empty($token)) {
```
Comment (batopa) [inline plugins/BEdita/API/src/Middleware/TokenMiddleware.php]: We could add a check on `$request->getAttribute(static::PAYLOAD_REQUEST_ATTRIBUTE)` and use that payload if already present.

File: plugins/BEdita/API/src/Controller/LoginController.php [hunk 4]
```diff
@@ -126,6 +127,28 @@ public function login(): void
...
+        if (!empty($this->request->getData('grant_type'))) {
+            return;
+        }
+
+        $data = $this->request->getData();
+        if (empty($data)) {
+            $this->request = $this->request->withData('grant_type', 'refresh_token');
+        } elseif (!empty($data['username']) && !empty($data['password'])) {
+            $this->request = $this->request->withData('grant_type', 'password');
+        } elseif (!empty(['client_id'])) {
```
Comment (batopa) [inline plugins/BEdita/API/src/Controller/LoginController.php]:
```suggestion
        } elseif (!empty($data['client_id'])) {
```

File: plugins/BEdita/API/src/Middleware/TokenMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+{
+    use InstanceConfigTrait;
+
+    /**
+     * Default config for this object.
+     *
+     * - `header` The header where the token is stored. Defaults to `'Authorization'`.
+     * - `headerPrefix` The prefix to the token in header. Defaults to `'Bearer'`.
+     * - `queryParam` The query parameter where the token is passed as a fallback. Defaults to `'token'`.
+     * - `allowedAlgorithms` List of supported verification algorithms. Defaults to `['HS256']`.
```
Comment (batopa) [inline plugins/BEdita/API/src/Middleware/TokenMiddleware.php]: `allowedAlgorithms` doesn't exist in default conf.

Missing documentation for `apiKey` and `clientAuth` keys.

File: plugins/BEdita/API/src/Middleware/TokenMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+     */
+    protected function readApplication(array $payload, ServerRequestInterface $request): void
+    {
+        $app = (array)Hash::get($payload, 'app');
+        if (!empty($app) && !empty($app['id'])) {
+            $application = new Application($app);
+            CurrentApplication::setApplication($application);
+        } else {
+            $this->applicationFromApiKey($request);
+        }
```
Comment (batopa) [inline plugins/BEdita/API/src/Middleware/TokenMiddleware.php]:
```suggestion
            CurrentApplication::setApplication($application);
            
            return;
        }
        
        $this->applicationFromApiKey($request);
```

File: plugins/BEdita/API/src/Utility/JWTHandler.php [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        // Common claims
+        $claims = [
+            'iss' => Router::fullBaseUrl(),
+            'iat' => time(),
+            'nbf' => time(),
+            'exp' => strtotime($duration),
+            'app' => static::applicationData(),
+        ];
+        // Access token payload
+        $payload = $user + $claims;
```
Comment (batopa) [inline plugins/BEdita/API/src/Utility/JWTHandler.php]: Maybe better `$payload = $claims + $user;` unless we want that `$user` data win against defined `$claims`.

File: plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+        $this->assertResponseCode(200);
+        $this->assertContentType('application/vnd.api+json');
+        $this->assertResponseNotEmpty();
+        $body = json_decode((string)$this->_response->getBody(), true);
+
+        $jwt = Hash::get($body, 'meta.jwt');
+        static::assertNotNull($jwt);
+        $renew = Hash::get($body, 'meta.renew');
+        static::assertNotNull($jwt);
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php]:
```suggestion
        static::assertNotNull($renew);
```

File: plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php]:
```suggestion
        static::assertNotNull($renew);
```

File: plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        $this->configRequest(compact('headers'));
+        $this->post('/auth', json_encode(['grant_type' => 'refresh_token']));
+
+        $this->assertResponseCode(200);
+        $body = json_decode((string)$this->_response->getBody(), true);
+
+        $jwt = Hash::get($body, 'meta.jwt');
+        static::assertNotNull($jwt);
+        $renew = Hash::get($body, 'meta.renew');
+        static::assertNotNull($jwt);
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/AuthenticationTest.php]:
```suggestion
        static::assertNotNull($renew);
```

File: plugins/BEdita/API/src/Auth/JwtAuthenticate.php [hunk 11]
```diff
@@ -129,93 +98,78 @@ public function authenticate(ServerRequest $request, Response $response)
...
+    protected function handleRefreshToken(ServerRequest $request)
     {
-        $token = $this->getToken($request);
-        if ($token) {
-            return $this->payload = $this->decode($token, $request);
+        $this->checkAudience($request);
+
+        if ($this->getConfig('authenticate') && !empty($this->payload['sub'])) {
+            return $this->_findUser($this->payload['sub']);
+        } elseif (empty($this->payload['sub']) && !empty($this->payload['app'])) {
```
Comment (batopa) [inline plugins/BEdita/API/src/Auth/JwtAuthenticate.php]:
```suggestion
        }
        
        if (empty($this->payload['sub']) && !empty($this->payload['app'])) {
```

File: plugins/BEdita/API/tests/TestCase/Auth/JwtAuthenticateTest.php [hunk 12]
```diff
@@ -52,228 +54,206 @@ class JwtAuthenticateTest extends TestCase
...
+        $auth = new JwtAuthenticate(new ComponentRegistry($controller), []);
+        $request = $request->withAttribute(TokenMiddleware::PAYLOAD_REQUEST_ATTRIBUTE, $data)
+            ->withData('grant_type', 'refresh_token');
+        $result = $auth->authenticate($request, new Response());
+
+        static::assertFalse($result);
+    }
+
+    /**
+     * Undocumented function
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Auth/JwtAuthenticateTest.php]: Undocumented.

File: plugins/BEdita/API/tests/TestCase/Controller/LoginControllerTest.php [hunk 13]
```diff
@@ -134,6 +139,138 @@ public function testSuccessfulRenew(array $meta)
...
+                ['client_id' => API_KEY],
+            ],
+        ];
+    }
+
+    /**
+     * Test `setGrantType()` method
+     *
+     * @param string $expected
+     * @param array $post
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Controller/LoginControllerTest.php]: Missing params description.

File: plugins/BEdita/API/tests/TestCase/Controller/LoginControllerTest.php [hunk 14]
```diff
@@ -134,6 +139,138 @@ public function testSuccessfulRenew(array $meta)
...
+                ['client_id' => 'gustavo'],
+            ]
+        ];
+    }
+
+    /**
+     * Test `checkClientCredentials()` method
+     *
+     * @param string $expected
+     * @param array $post
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Controller/LoginControllerTest.php]: Missing params description.

File: plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+                ],
+                [
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * Test middleware invocation.
+     *
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php:129]: Missing function params

File: plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testGetApplicationDefault()
+    {
+        $this->expectException(ForbiddenException::class);
+        $this->expectExceptionMessage('Missing API key');
+
+        Configure::delete('Security.blockAnonymousApps');
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php:188]: Do we need to restore previously value of `'Security.blockAnonymousApps'` at the end of test?

File: plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+     *
+     * @return void
+     *
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testClientCredentialsRequest()
+    {
+        CurrentApplication::setApplication(null);
+        Configure::delete('Security.blockAnonymousApps');
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php:209]: Do we need to restore previously value of `'Security.blockAnonymousApps'` at the end of test?

File: plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+                    'detail' => __d('bedita', 'Provided token has expired'),
+                    'code' => 'be_token_expired',
+                ]),
+                $expiredToken,
+            ],
+        ];
+    }
+
+    /**
+     * Test `getUser` method.
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php]:
```suggestion
     * Test `decode` method.
```

File: plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+        static::assertArrayHasKey('jwt', $tokens);
+        static::assertArrayHasKey('renew', $tokens);
+
+        $jwt = (array)JWT::decode($tokens['jwt'], Security::getSalt(), ['HS256']);
+        $expected = $user + [
+            'iss' => Router::fullBaseUrl(),
+            'app' => null,
+        ];
+        unset($jwt['iat'], $jwt['nbf'], $jwt['exp']);
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php]: maybe we could test the array keys presence before unset them.

File: plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        static::assertEquals($expected, $jwt);
+
+        $renew = (array)JWT::decode($tokens['renew'], Security::getSalt(), ['HS256']);
+        $expected = [
+            'iss' => Router::fullBaseUrl(),
+            'app' => null,
+            'sub' => 99,
+            'aud' => 'http://api.example.org',
+        ];
+        unset($renew['iat'], $renew['nbf'], $renew['exp']);
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Utility/JWTHandlerTest.php:132]: maybe we could test the array keys presence before unset them.

**Review by batopa — CHANGES_REQUESTED** [review]

File: plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php (hunk 16 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php:188]: In previous tests we did not keep state of this configuration. We can assume it's deleted or not defined as default, 
Here we just make sure it is deleted.

File: plugins/BEdita/API/src/Middleware/TokenMiddleware.php (hunk 5 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/src/Middleware/TokenMiddleware.php]: done

File: plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php (hunk 15 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php:129]: done

**Review by batopa — APPROVED** [review]

**Pull request merged**